### PR TITLE
Implement suite stalled handler.

### LINKED ISSUE #1286: suite stalled warning or notification
A suite is using a runahead limit because the suite cycles have no dependancies on previous cycles, but need to be limited due to large data volumes.
It would be helpful that if the suite reaches its runahead limit because of an earlier task has failed that a warning message is recorded in the stderr (and mail is sent as an option) to aid the user to identify why the suite is `waiting`.

### PR DESCRIPTION
**Pull request by arjclark**
Close #1286 

First pass attempt at implementing a suite stalled handler.

This takes a look at the current task pool and if no tasks are active, submitting (ready), queued or awaiting a clock trigger then the suite is deemed to have stalled. Given the way the scheduler loop runs, the "am I stalled" check needs to indicate a stalled state for two consecutive loops in order to fire off the stalled handler/abort on stalled. This captures where a state has updated to succeeded/failed but the follow on task has yet to trigger (happening in the next scheduler loop).

@hjoliver - can you take an initial look through and see if you think I've missed anything? If you're happy then I'll start writing tests, documenting etc.

### CODE REVIEW

**Comment by hjoliver** [comment]
@arjclark - this looks good, and the logic is simpler than my suggestion here https://github.com/cylc/cylc/issues/1286#issuecomment-70734218 :+1:

**Comment by arjclark** [comment]
@hjoliver - I think this now has everything needed, please review 1.
@matthewrmshin - please review 2

File: doc/suiterc.tex [hunk 1]
```diff
@@ -339,6 +340,19 @@ \subsection{[cylc]}
     \item {\em default:} True
 \end{myitemize}
 
+\paragraph[abort on stalled]{[cylc] \textrightarrow [[event hooks]] \textrightarrow abort on stalled}
+
+If a this is set is set to True this will cause the suite to abort with
```
Comment (matthewrmshin) [inline doc/suiterc.tex]: *If this is set to* ...?

Comment (arjclark) [inline doc/suiterc.tex]: corrected

File: lib/cylc/task_pool.py [hunk 2]
```diff
@@ -756,6 +756,20 @@ def has_unkillable_tasks_only(self):
                     return False
         return True
 
+    def pool_is_stalled(self):
+        """Return True if no active, queued or clock trigger awaiting tasks"""
+        for itask in self.get_tasks():
+            if (itask.state.status in [TASK_STATUS_QUEUED,
+                                       TASK_STATUS_READY] or
+               itask.state.status in TASK_STATUSES_ACTIVE):
```
Comment (matthewrmshin) [inline lib/cylc/task_pool.py]: Is it just Github or does this indentation look funny?

Comment (arjclark) [inline lib/cylc/task_pool.py]: The indentation *is* funny in order to pass the pep8 test - complains if the row doesn't start exactly there...

Comment (matthewrmshin) [inline lib/cylc/task_pool.py]: It is probably best to use an extra 4 spaces?

**Comment by matthewrmshin** [comment]
Looks OK 2. Test battery OK. Is it worth having a test for *abort on stalled*? (given that it is one of the main use cases.)

**Comment by arjclark** [comment]
@matthewrmshin - feedback addressed.

**Comment by matthewrmshin** [comment]
Looks OK 2. Tests OK in my environment.

File: doc/siterc.tex [hunk 3]
```diff
@@ -858,6 +858,10 @@ \subsection{[cylc]}
 
 \subparagraph[abort on timeout]{[cylc] \textrightarrow [[event hooks]] \textrightarrow abort on timeout}
 
+\subparagraph[startup handler]{[cylc] \textrightarrow [[event hooks]] \textrightarrow stalled handler}
+
```
Comment (hjoliver) [inline doc/siterc.tex]: The first bit here should be "stalled handler".

File: doc/siterc.tex [hunk 4]
```diff
@@ -858,6 +858,10 @@ \subsection{[cylc]}
...
+\subparagraph[abort on timeout]{[cylc] \textrightarrow [[event hooks]] \textrightarrow abort on stalled}
+
```
Comment (hjoliver) [inline doc/siterc.tex]: And the first bit here should be "abort on stalled".

File: doc/suiterc.tex [hunk 5]
```diff
@@ -339,6 +340,19 @@ \subsection{[cylc]}
     \item {\em default:} True
 \end{myitemize}
 
+\paragraph[abort on stalled]{[cylc] \textrightarrow [[event hooks]] \textrightarrow abort on stalled}
+
+If a this is set to True this will cause the suite to abort with error status
```
Comment (hjoliver) [inline doc/suiterc.tex]: "If a this"

File: doc/suiterc.tex [hunk 6]
```diff
@@ -339,6 +340,19 @@ \subsection{[cylc]}
...
+if the suite stalls. A suite is considered "stalled" if there are no active,
```
Comment (hjoliver) [inline doc/suiterc.tex]: "if it stalls"  (too many "suite"s)

**Comment by hjoliver** [comment]
Review 1: good, just a few doc typos to be fixed.

**Comment by arjclark** [comment]
@hjoliver - corrections made.

**Pull request merged**